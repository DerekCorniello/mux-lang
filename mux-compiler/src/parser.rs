use crate::lexer::Token;

#[derive(Debug, Clone)]
pub enum AstNode {
    Declaration(DeclarationNode),
    Statement(StatementNode),
    Expression(ExpressionNode),
    Type(TypeNode),
}

#[derive(Debug, Clone)]
pub enum DeclarationNode {
    Function {
        name: String,
        type_params: Vec<(String, Vec<TraitBound>)>,
        params: Vec<Param>,
        return_type: TypeNode,
        body: Vec<StatementNode>,
    },
    Class {
        name: String,
        type_params: Vec<(String, Vec<TraitBound>)>,
        traits: Vec<TraitRef>,
        fields: Vec<Field>,
        methods: Vec<FunctionNode>,
    },
    Interface {
        name: String,
        type_params: Vec<(String, Vec<TraitBound>)>,
        methods: Vec<FunctionNode>,
    },
    Enum {
        name: String,
        type_params: Vec<(String, Vec<TraitBound>)>,
        variants: Vec<EnumVariant>,
    },
    TypeAlias {
        name: String,
        type_params: Vec<(String, Vec<TraitBound>)>,
        target: TypeNode,
    },
    Const {
        name: String,
        type_: Option<TypeNode>,
        value: ExpressionNode,
    },
    Let {
        name: String,
        type_: Option<TypeNode>,
        value: ExpressionNode,
    },
}

#[derive(Debug, Clone)]
pub enum StatementNode {
    Let {
        name: String,
        type_: Option<TypeNode>,
        value: ExpressionNode,
    },
    Return(Option<ExpressionNode>),
    Expression(ExpressionNode),
    If {
        cond: ExpressionNode,
        then_block: Vec<StatementNode>,
        else_block: Option<Vec<StatementNode>>,
    },
    While {
        cond: ExpressionNode,
        body: Vec<StatementNode>,
    },
    For {
        var: String,
        iter: ExpressionNode,
        body: Vec<StatementNode>,
    },
    Break,
    Continue,
    Block(Vec<StatementNode>),
}

#[derive(Debug, Clone)]
pub enum ExpressionNode {
    Literal(LiteralNode),
    Identifier(String),
    Binary {
        left: Box<ExpressionNode>,
        op: BinaryOp,
        right: Box<ExpressionNode>,
    },
    Unary {
        op: UnaryOp,
        expr: Box<ExpressionNode>,
    },
    Call {
        func: Box<ExpressionNode>,
        args: Vec<ExpressionNode>,
    },
    FieldAccess {
        expr: Box<ExpressionNode>,
        field: String,
    },
    ArrayAccess {
        expr: Box<ExpressionNode>,
        index: Box<ExpressionNode>,
    },
    Lambda {
        params: Vec<Param>,
        body: Vec<StatementNode>,
    },
    IfExpr {
        cond: Box<ExpressionNode>,
        then_block: Vec<StatementNode>,
        else_block: Option<Vec<StatementNode>>,
    },
    Match {
        expr: Box<ExpressionNode>,
        arms: Vec<MatchArm>,
    },
}

#[derive(Debug, Clone)]
pub enum PrimitiveType {
    Int,
    Float,
    Bool,
    Char,
    Str,
}

#[derive(Debug, Clone)]
pub enum TypeNode {
    Primitive(PrimitiveType),
    Named(String),
    Generic {
        base: String,
        args: Vec<TypeNode>,
        bounds: Vec<TraitBound>,
    },
    List(Box<TypeNode>),
    Map(Box<TypeNode>, Box<TypeNode>),
    Array(Box<TypeNode>, usize),
    Tuple(Vec<TypeNode>),
    Optional(Box<TypeNode>),
    Result(Box<TypeNode>, Box<TypeNode>),
    Function {
        params: Vec<TypeNode>,
        returns: Box<TypeNode>,
    },
    Alias(String),
}

#[derive(Debug, Clone)]
pub struct TraitBound {
    pub name: String,
    pub type_params: Vec<TypeNode>,
}

#[derive(Debug, Clone)]
pub enum LiteralNode {
    Number(f64),
    String(String),
    Boolean(bool),
    Char(char),
}

#[derive(Debug, Clone)]
pub struct Param {
    pub name: String,
    pub type_: TypeNode,
}

#[derive(Debug, Clone)]
pub struct Field {
    pub name: String,
    pub type_: TypeNode,
}

#[derive(Debug, Clone)]
pub struct FunctionNode {
    pub name: String,
    pub type_params: Vec<(String, Vec<TraitBound>)>,
    pub params: Vec<Param>,
    pub return_type: TypeNode,
    pub body: Vec<StatementNode>,
}

#[derive(Debug, Clone)]
pub struct TraitRef {
    pub name: String,
    pub type_args: Vec<TypeNode>,
}

#[derive(Debug, Clone)]
pub struct EnumVariant {
    pub name: String,
    pub data: Option<Vec<TypeNode>>,
}

#[derive(Debug, Clone)]
pub struct MatchArm {
    pub pattern: PatternNode,
    pub body: Vec<StatementNode>,
}

#[derive(Debug, Clone)]
pub enum PatternNode {
    Literal(LiteralNode),
    Identifier(String),
    Wildcard, // `_`
    Tuple(Vec<PatternNode>),
    EnumVariant {
        name: String,
        args: Vec<PatternNode>,
    },
}

#[derive(Debug, Clone)]
pub enum BinaryOp {
    Add,
    Sub,
    Mul,
    Div,
    Mod,
    Eq,
    Neq,
    Lt,
    Gt,
    Leq,
    Geq,
    And,
    Or,
}

#[derive(Debug, Clone)]
pub enum UnaryOp {
    Neg,
    Not,
}

fn parse_from_tokens(tokens: Vec<Token>) -> Vec<AstNode> {
    let parsed_result: Vec<AstNode> = Vec::new();

    parsed_result
}
