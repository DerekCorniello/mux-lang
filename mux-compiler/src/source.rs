use std::io::{Error, ErrorKind};
use std::path::Path;

pub struct Source {
    pub input: String,
    pub pos: usize, // position in bytes
    pub line: usize,
    pub col: usize,
}

impl Source {
    pub fn new(file_path: &str) -> std::io::Result<Self> {
        if Path::new(file_path).exists() {
            // if file doesnt exist or there is an issue opening
            // it, then it will return the error with the ? op
            let input = std::fs::read_to_string(file_path)?;
            return Ok(Self {
                input,
                pos: 0,
                line: 1,
                col: 1,
            });
        }
        Err(Error::new(ErrorKind::NotFound, "File does not exist"))
    }

    #[allow(dead_code)]
    pub fn from_test_str(string: &str) -> Source {
        Source {
            input: string.to_string(),
            pos: 0,
            line: 1,
            col: 1,
        }
    }

    // get next char and move pos
    pub fn next_char(&mut self) -> Option<char> {
        if self.pos >= self.input.len() {
            return None;
        }
        let ch = self.input[self.pos..].chars().next()?;
        let char_len = ch.len_utf8();
        self.pos += char_len;

        if ch == '\n' {
            self.line += 1;
            self.col = 1;
        } else {
            // for multi-byte characters, we need to count the number of columns
            // based on the character's width in the terminal
            self.col += unicode_width::UnicodeWidthChar::width(ch).unwrap_or(1);
        }

        Some(ch)
    }

    // gets next char without consuming
    pub fn peek(&self) -> Option<char> {
        if self.pos >= self.input.len() {
            return None;
        }
        self.input[self.pos..].chars().next()
    }
    
    // consumes characters until the specified stop character is found. the stop
    // character is not consumed and not included in the returned string. leading
    // and trailing whitespace is trimmed from the result.
    pub fn consume_until(&mut self, stop: char) -> String {
        let mut buf = String::new();
        while let Some(c) = self.peek() {
            if c == stop {
                break;
            }
            buf.push(c);
            self.next_char();
        }
        buf.trim().to_string()
    }
    
    // consumes characters until the end of a multiline comment ("*/") is found. the
    // comment closer is consumed but not included in the returned string. the opening
    // "/*" should already be consumed before calling this.
    pub fn consume_multiline_comment(&mut self) -> String {
        let mut buf = String::new();
        
        while let Some(c) = self.next_char() {
            if c == '*' && self.peek() == Some('/') {
                self.next_char(); // consume the '/'
                break;
            }
            buf.push(c);
        }
        
        buf.trim().to_string()
    }
}

#[cfg(test)]
mod tests {
    use super::Source;

    #[test]
    fn test_next_char_basic() {
        let mut src = Source {
            input: "abc\n".to_string(),
            pos: 0,
            line: 1,
            col: 1,
        };

        assert_eq!(src.next_char(), Some('a'));
        assert_eq!(src.line, 1);
        assert_eq!(src.col, 2);

        assert_eq!(src.next_char(), Some('b'));
        assert_eq!(src.line, 1);
        assert_eq!(src.col, 3);

        assert_eq!(src.next_char(), Some('c'));
        assert_eq!(src.line, 1);
        assert_eq!(src.col, 4);

        assert_eq!(src.next_char(), Some('\n'));
        assert_eq!(src.line, 2);
        assert_eq!(src.col, 1);

        assert_eq!(src.next_char(), None);
    }

    #[test]
    fn test_peek() {
        let mut src = Source {
            input: "xy".to_string(),
            pos: 0,
            line: 1,
            col: 1,
        };

        assert_eq!(src.peek(), Some('x')); // peek doesn't advance pos
        assert_eq!(src.line, 1);
        assert_eq!(src.col, 1); // position unchanged after peek

        assert_eq!(src.next_char(), Some('x'));
        assert_eq!(src.line, 1);
        assert_eq!(src.col, 2);

        assert_eq!(src.peek(), Some('y'));
        assert_eq!(src.line, 1);
        assert_eq!(src.col, 2); // position unchanged after peek

        assert_eq!(src.next_char(), Some('y'));
        assert_eq!(src.line, 1);
        assert_eq!(src.col, 3);

        assert_eq!(src.peek(), None);
    }

    #[test]
    fn test_multiline_positioning() {
        let mut src = Source::from_test_str("hello\nworld\n");

        // test that positions track correctly across multiple lines
        for expected_char in "hello".chars() {
            assert_eq!(src.next_char(), Some(expected_char));
        }
        assert_eq!(src.line, 1);
        assert_eq!(src.col, 6);

        assert_eq!(src.next_char(), Some('\n')); // newline
        assert_eq!(src.line, 2);
        assert_eq!(src.col, 1);

        for expected_char in "world".chars() {
            assert_eq!(src.next_char(), Some(expected_char));
        }
        assert_eq!(src.line, 2);
        assert_eq!(src.col, 6);
    }

    #[test]
    fn test_consume_until() {
        let mut src = Source::from_test_str("hello world");

        let result = src.consume_until(' ');
        assert_eq!(result, "hello");
        assert_eq!(src.line, 1);
        assert_eq!(src.col, 6); // position should be at the space
        assert_eq!(src.next_char(), Some(' ')); // consume the space
        assert_eq!(src.next_char(), Some('w')); // next character should be 'w'
    }

    #[test]
    fn test_consume_multiline_comment() {
        // the multiline comment will already pass in the text
        // without the opening comment, thus the missing /*
        let mut src = Source::from_test_str(" hello\nworld*/");

        let result = src.consume_multiline_comment();
        assert_eq!(result, "hello\nworld");
        assert_eq!(src.line, 2); // should be on line 2 after the newline
        assert_eq!(src.col, 8); // after "world */"
    }

    #[test]
    fn test_unicode_handling() {
        let mut src = Source::from_test_str("„Åì„Çì„Å´„Å°„ÅØ");

        // First character '„Åì' is a full-width character (2 columns in terminal)
        let ch = src.next_char();
        println!("First char: {:?}, col: {}", ch, src.col);
        assert_eq!(ch, Some('„Åì'));
        assert_eq!(src.col, 3); // 1 (start) + 2 (full-width char) = 3

        // Second character '„Çì' is also a full-width character
        let ch = src.next_char();
        println!("Second char: {:?}, col: {}", ch, src.col);
        assert_eq!(ch, Some('„Çì'));
        assert_eq!(src.col, 5); // 3 + 2 = 5

        // Third character '„Å´' is also a full-width character
        let ch = src.next_char();
        println!("Third char: {:?}, col: {}", ch, src.col);
        assert_eq!(ch, Some('„Å´'));
        assert_eq!(src.col, 7); // 5 + 2 = 7
        
        // Fourth character '„Å°' is also a full-width character
        let ch = src.next_char();
        println!("Fourth char: {:?}, col: {}", ch, src.col);
        assert_eq!(ch, Some('„Å°'));
        assert_eq!(src.col, 9); // 7 + 2 = 9
        
        // Fifth character '„ÅØ' is also a full-width character
        let ch = src.next_char();
        println!("Fifth char: {:?}, col: {}", ch, src.col);
        assert_eq!(ch, Some('„ÅØ'));
        assert_eq!(src.col, 11); // 9 + 2 = 11
        
        // Test with emoji (typically 4 bytes, but displayed as one column)
        let mut src = Source::from_test_str("hello üåü");
        // Skip "hello " (6 characters including space)
        for _ in 0..6 {
            src.next_char();
        }
        
        // The star emoji is a double-width character
        // We've consumed "hello " (6 characters, columns 1-6)
        assert_eq!(src.col, 7); // After 6 characters, we're at column 7
        assert_eq!(src.next_char(), Some('üåü'));
        assert_eq!(src.col, 9); // Should increment by 2 for the emoji
    }

    #[test]
    fn test_from_test_str_type_consistency() {
        let src = Source::from_test_str("test");
        assert_eq!(src.input, "test".to_string());
    }
}
