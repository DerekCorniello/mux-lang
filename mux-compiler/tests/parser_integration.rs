use insta::assert_debug_snapshot;
use mux_compiler::lexer::Lexer;
use mux_compiler::parser::Parser;
use mux_compiler::source::Source;
use std::path::Path;

fn parse_file_to_ast(test_file: &str) -> String {
    let mut src = Source::new(test_file)
        .unwrap_or_else(|_| panic!("Failed to open source file: {}", test_file));

    let mut lexer = Lexer::new(&mut src);
    let tokens = lexer.lex_all()
        .unwrap_or_else(|e| panic!("Lexing failed: {}", e));
        
    let mut parser = Parser::new(&tokens);
    let result = parser.parse()
        .unwrap_or_else(|(_, errors)| {
            panic!("Parsing failed with errors: {:#?}", errors);
        });
    
    // Convert the AST to a nicely formatted string for snapshots
    let mut output = String::new();
    for node in result {
        output.push_str(&format!("{:#?}\n\n", node));
    }
    output
}

#[test]
fn test_control_flow_parsing() {
    let test_file = "../test_scripts/control_flow.mux";
    
    // Get the AST as a string
    let ast_string = parse_file_to_ast(test_file);
    
    // Create a snapshot name based on the test file name
    let snapshot_name = Path::new(test_file)
        .file_stem()
        .unwrap()
        .to_str()
        .unwrap();
    
    // Assert the snapshot
    assert_debug_snapshot!(snapshot_name, ast_string);
}
