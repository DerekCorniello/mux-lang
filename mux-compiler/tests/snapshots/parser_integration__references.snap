---
source: tests/parser_integration.rs
expression: ast_string
---
"Statement(\n    StatementNode {\n        kind: Import {\n            module_path: \"std.print\",\n            alias: None,\n        },\n        span: Span {\n            row_start: 1,\n            row_end: Some(\n                1,\n            ),\n            col_start: 1,\n            col_end: Some(\n                7,\n            ),\n        },\n    },\n)\n\nStatement(\n    StatementNode {\n        kind: TypedDecl(\n            \"val\",\n            TypeNode {\n                kind: Primitive(\n                    Int,\n                ),\n                span: Span {\n                    row_start: 4,\n                    row_end: Some(\n                        4,\n                    ),\n                    col_start: 1,\n                    col_end: Some(\n                        4,\n                    ),\n                },\n            },\n            ExpressionNode {\n                kind: Literal(\n                    Integer(\n                        10,\n                    ),\n                ),\n                span: Span {\n                    row_start: 4,\n                    row_end: Some(\n                        4,\n                    ),\n                    col_start: 11,\n                    col_end: Some(\n                        13,\n                    ),\n                },\n            },\n        ),\n        span: Span {\n            row_start: 4,\n            row_end: Some(\n                4,\n            ),\n            col_start: 1,\n            col_end: Some(\n                13,\n            ),\n        },\n    },\n)\n\nStatement(\n    StatementNode {\n        kind: AutoDecl(\n            \"r\",\n            TypeNode {\n                kind: Auto,\n                span: Span {\n                    row_start: 5,\n                    row_end: Some(\n                        5,\n                    ),\n                    col_start: 6,\n                    col_end: Some(\n                        7,\n                    ),\n                },\n            },\n            ExpressionNode {\n                kind: Unary {\n                    op: Ref,\n                    expr: ExpressionNode {\n                        kind: Identifier(\n                            \"val\",\n                        ),\n                        span: Span {\n                            row_start: 5,\n                            row_end: Some(\n                                5,\n                            ),\n                            col_start: 11,\n                            col_end: Some(\n                                14,\n                            ),\n                        },\n                    },\n                    postfix: false,\n                },\n                span: Span {\n                    row_start: 5,\n                    row_end: Some(\n                        5,\n                    ),\n                    col_start: 10,\n                    col_end: Some(\n                        14,\n                    ),\n                },\n            },\n        ),\n        span: Span {\n            row_start: 5,\n            row_end: Some(\n                5,\n            ),\n            col_start: 1,\n            col_end: Some(\n                14,\n            ),\n        },\n    },\n)\n\nStatement(\n    StatementNode {\n        kind: Expression(\n            ExpressionNode {\n                kind: Call {\n                    func: ExpressionNode {\n                        kind: Identifier(\n                            \"print\",\n                        ),\n                        span: Span {\n                            row_start: 6,\n                            row_end: Some(\n                                6,\n                            ),\n                            col_start: 1,\n                            col_end: Some(\n                                6,\n                            ),\n                        },\n                    },\n                    args: [\n                        ExpressionNode {\n                            kind: Identifier(\n                                \"r\",\n                            ),\n                            span: Span {\n                                row_start: 6,\n                                row_end: Some(\n                                    6,\n                                ),\n                                col_start: 7,\n                                col_end: Some(\n                                    8,\n                                ),\n                            },\n                        },\n                    ],\n                },\n                span: Span {\n                    row_start: 6,\n                    row_end: Some(\n                        6,\n                    ),\n                    col_start: 1,\n                    col_end: Some(\n                        6,\n                    ),\n                },\n            },\n        ),\n        span: Span {\n            row_start: 6,\n            row_end: Some(\n                6,\n            ),\n            col_start: 1,\n            col_end: Some(\n                6,\n            ),\n        },\n    },\n)\n\nStatement(\n    StatementNode {\n        kind: Expression(\n            ExpressionNode {\n                kind: Binary {\n                    left: ExpressionNode {\n                        kind: Identifier(\n                            \"r\",\n                        ),\n                        span: Span {\n                            row_start: 7,\n                            row_end: Some(\n                                7,\n                            ),\n                            col_start: 1,\n                            col_end: Some(\n                                2,\n                            ),\n                        },\n                    },\n                    op: Assign,\n                    right: ExpressionNode {\n                        kind: Literal(\n                            Integer(\n                                20,\n                            ),\n                        ),\n                        span: Span {\n                            row_start: 7,\n                            row_end: Some(\n                                7,\n                            ),\n                            col_start: 5,\n                            col_end: Some(\n                                7,\n                            ),\n                        },\n                    },\n                },\n                span: Span {\n                    row_start: 7,\n                    row_end: Some(\n                        7,\n                    ),\n                    col_start: 1,\n                    col_end: Some(\n                        7,\n                    ),\n                },\n            },\n        ),\n        span: Span {\n            row_start: 7,\n            row_end: Some(\n                7,\n            ),\n            col_start: 1,\n            col_end: Some(\n                7,\n            ),\n        },\n    },\n)\n\nStatement(\n    StatementNode {\n        kind: Expression(\n            ExpressionNode {\n                kind: Call {\n                    func: ExpressionNode {\n                        kind: Identifier(\n                            \"print\",\n                        ),\n                        span: Span {\n                            row_start: 8,\n                            row_end: Some(\n                                8,\n                            ),\n                            col_start: 1,\n                            col_end: Some(\n                                6,\n                            ),\n                        },\n                    },\n                    args: [\n                        ExpressionNode {\n                            kind: Identifier(\n                                \"val\",\n                            ),\n                            span: Span {\n                                row_start: 8,\n                                row_end: Some(\n                                    8,\n                                ),\n                                col_start: 7,\n                                col_end: Some(\n                                    10,\n                                ),\n                            },\n                        },\n                    ],\n                },\n                span: Span {\n                    row_start: 8,\n                    row_end: Some(\n                        8,\n                    ),\n                    col_start: 1,\n                    col_end: Some(\n                        6,\n                    ),\n                },\n            },\n        ),\n        span: Span {\n            row_start: 8,\n            row_end: Some(\n                8,\n            ),\n            col_start: 1,\n            col_end: Some(\n                6,\n            ),\n        },\n    },\n)\n\nStatement(\n    StatementNode {\n        kind: AutoDecl(\n            \"numbers\",\n            TypeNode {\n                kind: Auto,\n                span: Span {\n                    row_start: 11,\n                    row_end: Some(\n                        11,\n                    ),\n                    col_start: 6,\n                    col_end: Some(\n                        13,\n                    ),\n                },\n            },\n            ExpressionNode {\n                kind: ListLiteral(\n                    [\n                        ExpressionNode {\n                            kind: Literal(\n                                Integer(\n                                    1,\n                                ),\n                            ),\n                            span: Span {\n                                row_start: 11,\n                                row_end: Some(\n                                    11,\n                                ),\n                                col_start: 17,\n                                col_end: Some(\n                                    18,\n                                ),\n                            },\n                        },\n                        ExpressionNode {\n                            kind: Literal(\n                                Integer(\n                                    2,\n                                ),\n                            ),\n                            span: Span {\n                                row_start: 11,\n                                row_end: Some(\n                                    11,\n                                ),\n                                col_start: 20,\n                                col_end: Some(\n                                    21,\n                                ),\n                            },\n                        },\n                        ExpressionNode {\n                            kind: Literal(\n                                Integer(\n                                    3,\n                                ),\n                            ),\n                            span: Span {\n                                row_start: 11,\n                                row_end: Some(\n                                    11,\n                                ),\n                                col_start: 23,\n                                col_end: Some(\n                                    24,\n                                ),\n                            },\n                        },\n                        ExpressionNode {\n                            kind: Literal(\n                                Integer(\n                                    4,\n                                ),\n                            ),\n                            span: Span {\n                                row_start: 11,\n                                row_end: Some(\n                                    11,\n                                ),\n                                col_start: 26,\n                                col_end: Some(\n                                    27,\n                                ),\n                            },\n                        },\n                        ExpressionNode {\n                            kind: Literal(\n                                Integer(\n                                    5,\n                                ),\n                            ),\n                            span: Span {\n                                row_start: 11,\n                                row_end: Some(\n                                    11,\n                                ),\n                                col_start: 29,\n                                col_end: Some(\n                                    30,\n                                ),\n                            },\n                        },\n                    ],\n                ),\n                span: Span {\n                    row_start: 11,\n                    row_end: None,\n                    col_start: 16,\n                    col_end: None,\n                },\n            },\n        ),\n        span: Span {\n            row_start: 11,\n            row_end: Some(\n                11,\n            ),\n            col_start: 1,\n            col_end: Some(\n                5,\n            ),\n        },\n    },\n)\n\nStatement(\n    StatementNode {\n        kind: AutoDecl(\n            \"first\",\n            TypeNode {\n                kind: Auto,\n                span: Span {\n                    row_start: 12,\n                    row_end: Some(\n                        12,\n                    ),\n                    col_start: 6,\n                    col_end: Some(\n                        11,\n                    ),\n                },\n            },\n            ExpressionNode {\n                kind: Unary {\n                    op: Ref,\n                    expr: ExpressionNode {\n                        kind: ListAccess {\n                            expr: ExpressionNode {\n                                kind: Identifier(\n                                    \"numbers\",\n                                ),\n                                span: Span {\n                                    row_start: 12,\n                                    row_end: Some(\n                                        12,\n                                    ),\n                                    col_start: 15,\n                                    col_end: Some(\n                                        22,\n                                    ),\n                                },\n                            },\n                            index: ExpressionNode {\n                                kind: Literal(\n                                    Integer(\n                                        0,\n                                    ),\n                                ),\n                                span: Span {\n                                    row_start: 12,\n                                    row_end: Some(\n                                        12,\n                                    ),\n                                    col_start: 23,\n                                    col_end: Some(\n                                        24,\n                                    ),\n                                },\n                            },\n                        },\n                        span: Span {\n                            row_start: 12,\n                            row_end: Some(\n                                12,\n                            ),\n                            col_start: 15,\n                            col_end: Some(\n                                22,\n                            ),\n                        },\n                    },\n                    postfix: false,\n                },\n                span: Span {\n                    row_start: 12,\n                    row_end: Some(\n                        12,\n                    ),\n                    col_start: 14,\n                    col_end: Some(\n                        22,\n                    ),\n                },\n            },\n        ),\n        span: Span {\n            row_start: 12,\n            row_end: Some(\n                12,\n            ),\n            col_start: 1,\n            col_end: Some(\n                22,\n            ),\n        },\n    },\n)\n\nStatement(\n    StatementNode {\n        kind: Expression(\n            ExpressionNode {\n                kind: Call {\n                    func: ExpressionNode {\n                        kind: Identifier(\n                            \"print\",\n                        ),\n                        span: Span {\n                            row_start: 13,\n                            row_end: Some(\n                                13,\n                            ),\n                            col_start: 1,\n                            col_end: Some(\n                                6,\n                            ),\n                        },\n                    },\n                    args: [\n                        ExpressionNode {\n                            kind: Identifier(\n                                \"first\",\n                            ),\n                            span: Span {\n                                row_start: 13,\n                                row_end: Some(\n                                    13,\n                                ),\n                                col_start: 7,\n                                col_end: Some(\n                                    12,\n                                ),\n                            },\n                        },\n                    ],\n                },\n                span: Span {\n                    row_start: 13,\n                    row_end: Some(\n                        13,\n                    ),\n                    col_start: 1,\n                    col_end: Some(\n                        6,\n                    ),\n                },\n            },\n        ),\n        span: Span {\n            row_start: 13,\n            row_end: Some(\n                13,\n            ),\n            col_start: 1,\n            col_end: Some(\n                6,\n            ),\n        },\n    },\n)\n\nFunction(\n    FunctionNode {\n        name: \"update\",\n        type_params: [],\n        params: [\n            Param {\n                name: \"ref\",\n                type_: TypeNode {\n                    kind: Reference(\n                        TypeNode {\n                            kind: Primitive(\n                                Int,\n                            ),\n                            span: Span {\n                                row_start: 16,\n                                row_end: Some(\n                                    16,\n                                ),\n                                col_start: 14,\n                                col_end: Some(\n                                    17,\n                                ),\n                            },\n                        },\n                    ),\n                    span: Span {\n                        row_start: 16,\n                        row_end: Some(\n                            16,\n                        ),\n                        col_start: 13,\n                        col_end: Some(\n                            17,\n                        ),\n                    },\n                },\n            },\n        ],\n        return_type: TypeNode {\n            kind: Primitive(\n                Void,\n            ),\n            span: Span {\n                row_start: 16,\n                row_end: Some(\n                    16,\n                ),\n                col_start: 31,\n                col_end: Some(\n                    35,\n                ),\n            },\n        },\n        body: [\n            StatementNode {\n                kind: Expression(\n                    ExpressionNode {\n                        kind: Binary {\n                            left: ExpressionNode {\n                                kind: Identifier(\n                                    \"ref\",\n                                ),\n                                span: Span {\n                                    row_start: 17,\n                                    row_end: Some(\n                                        17,\n                                    ),\n                                    col_start: 5,\n                                    col_end: Some(\n                                        8,\n                                    ),\n                                },\n                            },\n                            op: Assign,\n                            right: ExpressionNode {\n                                kind: Binary {\n                                    left: ExpressionNode {\n                                        kind: Identifier(\n                                            \"ref\",\n                                        ),\n                                        span: Span {\n                                            row_start: 17,\n                                            row_end: Some(\n                                                17,\n                                            ),\n                                            col_start: 11,\n                                            col_end: Some(\n                                                14,\n                                            ),\n                                        },\n                                    },\n                                    op: Add,\n                                    right: ExpressionNode {\n                                        kind: Literal(\n                                            Integer(\n                                                1,\n                                            ),\n                                        ),\n                                        span: Span {\n                                            row_start: 17,\n                                            row_end: Some(\n                                                17,\n                                            ),\n                                            col_start: 17,\n                                            col_end: Some(\n                                                18,\n                                            ),\n                                        },\n                                    },\n                                },\n                                span: Span {\n                                    row_start: 17,\n                                    row_end: Some(\n                                        17,\n                                    ),\n                                    col_start: 11,\n                                    col_end: Some(\n                                        18,\n                                    ),\n                                },\n                            },\n                        },\n                        span: Span {\n                            row_start: 17,\n                            row_end: Some(\n                                17,\n                            ),\n                            col_start: 5,\n                            col_end: Some(\n                                18,\n                            ),\n                        },\n                    },\n                ),\n                span: Span {\n                    row_start: 17,\n                    row_end: Some(\n                        17,\n                    ),\n                    col_start: 5,\n                    col_end: Some(\n                        18,\n                    ),\n                },\n            },\n        ],\n        span: Span {\n            row_start: 16,\n            row_end: Some(\n                17,\n            ),\n            col_start: 1,\n            col_end: Some(\n                18,\n            ),\n        },\n    },\n)\n\nStatement(\n    StatementNode {\n        kind: Expression(\n            ExpressionNode {\n                kind: Call {\n                    func: ExpressionNode {\n                        kind: Identifier(\n                            \"update\",\n                        ),\n                        span: Span {\n                            row_start: 20,\n                            row_end: Some(\n                                20,\n                            ),\n                            col_start: 1,\n                            col_end: Some(\n                                7,\n                            ),\n                        },\n                    },\n                    args: [\n                        ExpressionNode {\n                            kind: Unary {\n                                op: Ref,\n                                expr: ExpressionNode {\n                                    kind: Identifier(\n                                        \"val\",\n                                    ),\n                                    span: Span {\n                                        row_start: 20,\n                                        row_end: Some(\n                                            20,\n                                        ),\n                                        col_start: 9,\n                                        col_end: Some(\n                                            12,\n                                        ),\n                                    },\n                                },\n                                postfix: false,\n                            },\n                            span: Span {\n                                row_start: 20,\n                                row_end: Some(\n                                    20,\n                                ),\n                                col_start: 8,\n                                col_end: Some(\n                                    12,\n                                ),\n                            },\n                        },\n                    ],\n                },\n                span: Span {\n                    row_start: 20,\n                    row_end: Some(\n                        20,\n                    ),\n                    col_start: 1,\n                    col_end: Some(\n                        7,\n                    ),\n                },\n            },\n        ),\n        span: Span {\n            row_start: 20,\n            row_end: Some(\n                20,\n            ),\n            col_start: 1,\n            col_end: Some(\n                7,\n            ),\n        },\n    },\n)\n\nStatement(\n    StatementNode {\n        kind: Expression(\n            ExpressionNode {\n                kind: Call {\n                    func: ExpressionNode {\n                        kind: Identifier(\n                            \"print\",\n                        ),\n                        span: Span {\n                            row_start: 21,\n                            row_end: Some(\n                                21,\n                            ),\n                            col_start: 1,\n                            col_end: Some(\n                                6,\n                            ),\n                        },\n                    },\n                    args: [\n                        ExpressionNode {\n                            kind: Identifier(\n                                \"val\",\n                            ),\n                            span: Span {\n                                row_start: 21,\n                                row_end: Some(\n                                    21,\n                                ),\n                                col_start: 7,\n                                col_end: Some(\n                                    10,\n                                ),\n                            },\n                        },\n                    ],\n                },\n                span: Span {\n                    row_start: 21,\n                    row_end: Some(\n                        21,\n                    ),\n                    col_start: 1,\n                    col_end: Some(\n                        6,\n                    ),\n                },\n            },\n        ),\n        span: Span {\n            row_start: 21,\n            row_end: Some(\n                21,\n            ),\n            col_start: 1,\n            col_end: Some(\n                6,\n            ),\n        },\n    },\n)\n\n"
